src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/UnmanagedTypeConstraintTests.vb

```diff
@@ -0,0 +1,1799 @@
+
+    void Tst()
+    {
+        this.M((RefS)default);
+    }
+}
+
+public ref struct RefS { }
+public ref struct RefG<T> { public T field; }
+public ref struct Ref { ref int field; }
+public ref struct StructWithIndirectRefField
+{
+    public Ref Field;
+}
+public ref struct StructWithIndirectRefField2
+{
+    public StructWithRefField<int> Field;
+}
+public ref struct StructWithRefField<T>
+{
+    public ref T RefField;
+}
+
+", parseOptions:=New CSharpParseOptions(CSharp.LanguageVersion.Latest), referencedAssemblies:=Basic.Reference.Assemblies.Net90.References.All).EmitToImageReference()
+
+            Dim source = "
+public class Program 
+    <System.Obsolete>
+    Shared Sub Main
+        Dim o As New Test
+        o.M(CType(Nothing, RefS))
+        o.M(CType(Nothing, RefG(Of String)))
+        o.M(CType(Nothing, Ref))
+        o.M(CType(Nothing, StructWithIndirectRefField))
+        o.M(CType(Nothing, StructWithIndirectRefField2))
+    End Sub
+End Class
+"
+
+            Dim compilation = CreateCompilation(source, references:={reference}, targetFramework:=TargetFramework.Net90)
+            AssertTheseDiagnostics(compilation, <expected><![CDATA[
+BC37332: Type argument 'RefG(Of String)' must be a non-nullable value type, along with all fields at any level of nesting, in order to use it as type parameter 'T'.
+        o.M(CType(Nothing, RefG(Of String)))
+          ~
+BC30656: Field 'field' is of an unsupported type.
+        o.M(CType(Nothing, Ref))
+        ~~~~~~~~~~~~~~~~~~~~~~~~
+BC30656: Field 'field' is of an unsupported type.
+        o.M(CType(Nothing, StructWithIndirectRefField))
+        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+BC30656: Field 'RefField' is of an unsupported type.
+        o.M(CType(Nothing, StructWithIndirectRefField2))
+        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+                                                ]]></expected>)
+        End Sub
+
+        <Fact>
+        Public Sub UnmanagedCheck_Class()
+            Dim reference = CreateCSharpCompilation("
+public class Test
+{
+    public void M<T>(T x) where T : unmanaged {}
+}
+", parseOptions:=New CSharpParseOptions(CSharp.LanguageVersion.Latest)).EmitToImageReference()
+
+            Dim source = "
+public class Program 
+    Shared Sub Main
+        Dim o As New Test
+        o.M(New Program())
+    End Sub
+End Class
+"
+
+            Dim compilation = CreateCompilation(source, references:={reference})
+            AssertTheseDiagnostics(compilation, <expected><![CDATA[
+BC32105: Type argument 'Program' does not satisfy the 'Structure' constraint for type parameter 'T'.
+        o.M(New Program())
+          ~
+BC37332: Type argument 'Program' must be a non-nullable value type, along with all fields at any level of nesting, in order to use it as type parameter 'T'.
+        o.M(New Program())
+          ~
+                                                ]]></expected>)
+        End Sub
```

Comment:
>End Sub [](http://example.com/codeflow?start=8&length=7)

Are we testing type parameters?
```
Sub M1(Of T1, T2 As Class, T3 As Structure)(t1 as T1, t2 As T2, t3 As T3)
    Dim o As New Test
    o.M(t1)
    o.M(t2)
    o.M(t3)
End Sub
```

Reply:
> Are we testing type parameters?

Type parameters are covered in `UnmanagedCheck_TypeParameter` and indirectly on several other tests.